sw/host/opentitanlib/src/transport/errors.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+/// Contains all the errors that any method on the `Transport` trait could generate.  This
+/// struct is serializable, such that it can be transmitted across a network for instance as
+/// part of the session proxy functionality.
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum TransportError {
+    #[error("USB device did not match")]
+    NoMatch,
+    #[error("Found no USB device")]
+    NoDevice,
+    #[error("Found multiple USB devices, use --serial")]
+    MultipleDevices,
+    #[error("USB error: {0}")]
+    UsbGenericError(String),
+    #[error("Error opening USB device: {0}")]
+    UsbOpenError(String),
+    #[error("Transport does not support {0} instance {1}")]
+    InvalidInstance(TransportInterfaceType, String),
+    #[error("Encountered non-unicode device path")]
+    UnicodePathError,
+    #[error("Error opening {0}: {1}")]
+    OpenError(String, String),
+    #[error("Error reading from {0}: {1}")]
+    ReadError(String, String),
+    #[error("FPGA programming failed: {0}")]
+    FpgaProgramFailed(String),
+    #[error("Invalid pin strapping name \"{0}\"")]
+    InvalidStrappingName(String),
+    #[error("Transport does not support the requested operation")]
+    UnsupportedOperation,
+    #[error("Error communicating with FTDI: {0}")]
+    FtdiError(String),
+    #[error("Error communicating with debugger: {0}")]
+    CommunicationError(String),
+
+    // Include sub-enums for the various sub-traits of Tranport.
+    #[error("GPIO error: {0}")]
```

Comment:
Does `#[error(from)]` make sense here? (Can't remember the precise spelling but you can ask thiserror to generate From impls).

Reply:
Nice.